sampling/speculative_sampling.py

```diff
@@ -1,185 +1,210 @@
+                n = prefix_len + i - 1
+                break
+
+            if verbose:
+                print(f"approx guess accepted {j[0]}: \033[31m{Decoder().decode(torch.tensor([j]))}\033[0m")
+
+            accepted_count += 1
+
+        # print(f"n : {n}, i : {i}, prefix_len + gamma - 1: {prefix_len + gamma - 1}")
+        assert n >= prefix_len - 1, f"n {n}, prefix_len {prefix_len}"
+        prefix = x[:, :n + 1]
+
+        approx_model_cache.rollback(n + 1)
+
+        assert approx_model_cache._prob_history.shape[
+                   -2] <= n + 1, f"approx_model prob list shape {approx_model_cache._prob_history.shape}, n {n}"
+
+        if n < prefix_len + gamma - 1:
+            # reject someone, sample from the pos n
+            t = sample(max_fn(target_model_cache._prob_history[:, n, :] - approx_model_cache._prob_history[:, n, :]))
+            if verbose:
+                print(f"target resamples at position {n}: \033[34m{Decoder().decode(t)}\033[0m")
+            resample_count += 1
+            target_model_cache.rollback(n + 1)
+        else:
+            # all approx model decoding accepted
+            assert n == target_model_cache._prob_history.shape[1] - 1
+            t = sample(target_model_cache._prob_history[:, -1, :])
+            if verbose:
+                print(f"target samples {n}: \033[35m{Decoder().decode(t)}\033[0m")
+            target_sample_count += 1
+            target_model_cache.rollback(n + 2)
+
+        prefix = torch.cat((prefix, t), dim=1)
+
+    if verbose:
+        print(
+            f"generated tokens numbers {prefix.shape[-1] - seq_len}, accepted_count {accepted_count}, target_sample_count {target_sample_count}, resample_count {resample_count}")
+    return prefix
+
+
+@torch.no_grad()
+def speculative_sampling_v2(prefix: torch.Tensor, approx_model: torch.nn.Module, target_model: torch.nn.Module,
+                            max_len: int, gamma: int = 4,
+                            temperature: float = 1, top_k: int = 0, top_p: float = 0,
+                            random_seed: int = None) -> torch.Tensor:
+    """
+    DeepMind version Speculative Sampling.
+    Accelerating Large Language Model Decoding with Speculative Sampling
+    https://arxiv.org/abs/2302.01318
+    No KV Cache Optimization
+
+    Args:
+        x (torch.Tensor): input sequence, (batch, prefix_seqlen), Note that the batch dim is always 1 now.
+        approx_model (torch.nn.Module): approx model, the small one
+        target_model (torch.nn.Module): target model, the large one
+        max_len (int): the max overall generated tokens number.
+        gamma (int): $\gamma$, the token number small model guesses.
+        temperature (float, optional): Defaults to 1.
+        top_k (int, optional): Defaults to 0.
+        top_p (float, optional): Defaults to 0.
+
+    Returns:
+        torch.Tensor: generated tokens (batch, target_seqlen)
+    """
+
+    def batch_padding(x, gamma):
+        PADDING_CODE = 2

```

Comment:
you can get the padding code from the model?

Reply:
padding_code = tokenizer.pad_token_id
like this way

Reply:
Revised in next version.